### ISSUE TITLE
Vcsim - CloneVapp_Task seems not implemented

### ISSUE THREAD

**Issue by gabriele-g-els** [vmware/govmomi#2028]
Hi, 
when I try to clone a VApp on vcsim I get the following response:

```
vcsim_1        | Request: <?xml version="1.0" encoding="utf-8"?>
                            <soap:Envelope xmlns:soap="http://schemas.xmlsoap.org/soap/envelope/" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:xsd="http://www.w3.org/2001/XMLSchema">
                                <soap:Body>
                                    <CloneVApp_Task xmlns="urn:vim25">
                                        <_this type="VirtualApp">resgroup-v639</_this>
                                        <name>1_1</name>
                                        <target type="ResourcePool">resgroup-186</target>
                                        <spec>
                                            <location type="Datastore">datastore-187</location>
                                            <host type="HostSystem">host-183</host>
                                            <vmFolder type="Folder">group-v635</vmFolder>
                                            <networkMapping>
                                                <source type="Network">network-307</source>
                                                <destination type="Network">network-2</destination>
                                            </networkMapping>
                                        </spec>
                                    </CloneVApp_Task>
                                </soap:Body>
                            </soap:Envelope>
vcsim_1        | 2020/06/22 14:55:55 VirtualApp:resgroup-v639 does not implement: CloneVApp_Task
vcsim_1        | Response: <?xml version="1.0" encoding="UTF-8"?>
vcsim_1        |                <soapenv:Envelope xmlns:soapenc="http://schemas.xmlsoap.org/soap/encoding/" xmlns:soapenv="http://schemas.xmlsoap.org/soap/envelope/" xmlns:xsd="http://www.w3.org/2001/XMLSchema" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
                                    <soapenv:Body>
                                        <soapenv:Fault>
                                            <faultcode>ServerFaultCode</faultcode>
                                            <faultstring>VirtualApp:resgroup-v639 does not implement: CloneVApp_Task</faultstring>
                                            <detail>
                                                <MethodNotFoundFault xmlns="urn:vim25" xsi:type="MethodNotFound">
                                                    <receiver type="VirtualApp">resgroup-v639</receiver>
                                                    <method>CloneVApp_Task</method>
                                                </MethodNotFoundFault>
                                            </detail>
                                        </soapenv:Fault>
                                    </soapenv:Body>
                                </soapenv:Envelope>
```
Can you help me on this issue?
Thanks

**Comment by oniric85** [comment]
Hello,
I'm interested on solving this issue too, is there some way we can help?

**Comment by dougm** [comment]
We need to implement the CloneVApp_Task method in vcsim.. the API doc doesn't provide any details. What does this method do, clone all VMs within a VApp?

**Comment by oniric85** [comment]
This is the only official resource I've been able to find googling: https://pubs.vmware.com/vsphere-50/index.jsp?topic=%2Fcom.vmware.wssdk.apiref.doc_50%2Fvim.VirtualApp.html
What it does is basically to clone all VMs within the source VApp to the target location (either a RP or another vApp) following passed clonespecs.

**Comment by dougm** [comment]
@gabriele-g-els @oniric85 initial support in #2111 , can you review / try it out?

**Comment by oniric85** [comment]
Hello @dougm thanks for the PR! I've tested it a bit this morning but I wasn't able to complete the test due to an error that surfaces when the NewVirtualMachine method inside virtual_machine.go is called:

`mkdir ds:/vmfs/volumes/5df2befa-f8eea24e-e7c9-002590bc5732/1_1: no such file or directory`

I've loaded an inventory with vcsim to perform the test. Do you have any suggestions? I think we are on the right path though!

**Comment by dougm** [comment]
Ah, if you're using save+load, there is a datastore path issue that needs fixing: https://github.com/vmware/govmomi/issues/1919#issuecomment-643081303

I'll try to get that fixed soon-ish.

**Comment by oniric85** [comment]
Yes, I've saved the inventory from a "real" vcenter using govc. But yeah, that is the exactly same error. In the meantime while you work on the other issue, Is another way I can help you test this?

**Comment by dougm** [comment]
You could try manually changing the saved .xml
In `*-Datastore-*.xml` replace 
`<url>ds:/vmfs/volumes/5df2befa-f8eea24e-e7c9-002590bc5732</url>` with 
`<url>/tmp/any-directory-vcsim-can-write-to</url>`

You'll need to create that local directory. Then restart `vcsim -load ...`.
I'm not certain if it'll all work with just that change though.

**Comment by oniric85** [comment]
Thanks for the help, I was able to get past the issue with file system. However I'm now struggling trying to solve another problem I've found. Here's the stacktrace

```
2020/09/15 11:10:42 http: panic serving 10.10.10.6:52886: runtime error: invalid memory address or nil pointer dereference
goroutine 25 [running]:
net/http.(*conn).serve.func1(0xc00066b400)
	/usr/local/go/src/net/http/server.go:1772 +0x139
panic(0xd74460, 0x18e3900)
	/usr/local/go/src/runtime/panic.go:975 +0x3e3
github.com/vmware/govmomi/simulator.(*VirtualMachine).configureDevice(0xc000653800, 0xc00153b600, 0xb, 0x10, 0xc00154cb90, 0x0, 0x0)
	/go/src/github.com/vmware/govmomi/simulator/virtual_machine.go:1048 +0x113f
github.com/vmware/govmomi/simulator.(*VirtualMachine).configureDevices(0xc000653800, 0xc00172c3a0, 0x7, 0x0)
	/go/src/github.com/vmware/govmomi/simulator/virtual_machine.go:1296 +0x3bd
github.com/vmware/govmomi/simulator.(*VirtualMachine).create(0xc000653800, 0xc00172c3a0, 0x0, 0xc0005b0624, 0xa)
	/go/src/github.com/vmware/govmomi/simulator/virtual_machine.go:924 +0x391
github.com/vmware/govmomi/simulator.(*createVM).Run(0xc00053e500, 0xc000d90d00, 0xc000d90d00, 0xc000c5f4a0, 0x2, 0x2)
	/go/src/github.com/vmware/govmomi/simulator/folder.go:387 +0x3b2
github.com/vmware/govmomi/simulator.(*Task).Run(0xc000d90d00, 0xc00053e500, 0xc000d90d00, 0x194b2c0, 0xc001687db8)
	/go/src/github.com/vmware/govmomi/simulator/task.go:88 +0x168
github.com/vmware/govmomi/simulator.(*Folder).CreateVMTask(0xc0008af520, 0xc00156ed80, 0xc00172c380, 0x7fe69c211410, 0xc0008af520)
	/go/src/github.com/vmware/govmomi/simulator/folder.go:440 +0x84
github.com/vmware/govmomi/simulator.(*VirtualMachine).CloneVMTask.func1(0xc000d90b60, 0x11571c0, 0xc000d90b60, 0xc000c5f440, 0x2)
	/go/src/github.com/vmware/govmomi/simulator/virtual_machine.go:1671 +0x506
github.com/vmware/govmomi/simulator.(*Task).Run(0xc000d90b60, 0xc001677000, 0xf90dab, 0x7, 0xc000bcc680)
	/go/src/github.com/vmware/govmomi/simulator/task.go:88 +0x168
github.com/vmware/govmomi/simulator.(*VirtualMachine).CloneVMTask(0xc001677000, 0xc00156ed80, 0xc00157dd40, 0xc001483618, 0x6)
	/go/src/github.com/vmware/govmomi/simulator/virtual_machine.go:1704 +0x521
github.com/vmware/govmomi/simulator.(*VirtualApp).CloneVAppTask.func1(0xc000d90820, 0x11571c0, 0xc000d90820, 0xc000c5f320, 0x2)
	/go/src/github.com/vmware/govmomi/simulator/resource_pool.go:394 +0x3b2
github.com/vmware/govmomi/simulator.(*Task).Run(0xc000d90820, 0xc000581500, 0xf920a1, 0x9, 0xc00051d080)
	/go/src/github.com/vmware/govmomi/simulator/task.go:88 +0x168
github.com/vmware/govmomi/simulator.(*VirtualApp).CloneVAppTask(0xc000581500, 0xc00156ed80, 0xc001542700, 0x0, 0x0)
	/go/src/github.com/vmware/govmomi/simulator/resource_pool.go:417 +0xa5
reflect.Value.call(0xf0bae0, 0xc000581500, 0x213, 0xf8f7d4, 0x4, 0xc000883c50, 0x2, 0x2, 0xd, 0xa, ...)
	/usr/local/go/src/reflect/value.go:460 +0x8ab
reflect.Value.Call(0xf0bae0, 0xc000581500, 0x213, 0xc000883c50, 0x2, 0x2, 0xc000d5a240, 0xa, 0xc000d5a260)
	/usr/local/go/src/reflect/value.go:321 +0xb4
github.com/vmware/govmomi/simulator.(*Service).call.func1()
	/go/src/github.com/vmware/govmomi/simulator/simulator.go:216 +0x64
github.com/vmware/govmomi/simulator.(*Registry).WithLock(0xc0002dc3c0, 0x1157260, 0xc000581500, 0xc0016896f0)
	/go/src/github.com/vmware/govmomi/simulator/registry.go:554 +0x47
github.com/vmware/govmomi/simulator.(*Service).call(0xc000e42480, 0xc00156ed80, 0xc000dac480, 0x600, 0xc000dac480)
	/go/src/github.com/vmware/govmomi/simulator/simulator.go:215 +0x5fc
github.com/vmware/govmomi/simulator.(*Service).ServeSDK(0xc000e42480, 0x1164a40, 0xc001542620, 0xc00153b200)
	/go/src/github.com/vmware/govmomi/simulator/simulator.go:472 +0x426
net/http.HandlerFunc.ServeHTTP(0xc000feb3f0, 0x1164a40, 0xc001542620, 0xc00153b200)
	/usr/local/go/src/net/http/server.go:2012 +0x44
net/http.(*ServeMux).ServeHTTP(0x1997b00, 0x1164a40, 0xc001542620, 0xc00153b200)
	/usr/local/go/src/net/http/server.go:2387 +0x1a5
net/http.serverHandler.ServeHTTP(0xc00144e700, 0x1164a40, 0xc001542620, 0xc00153b200)
	/usr/local/go/src/net/http/server.go:2807 +0xa3
net/http.(*conn).serve(0xc00066b400, 0x1166740, 0xc000a0f740)
	/usr/local/go/src/net/http/server.go:1895 +0x86c
created by net/http.(*Server).Serve
	/usr/local/go/src/net/http/server.go:2933 +0x35c
```
It appears to be an issue when the code is doing a lookup for a VM network in the datacenter Network folder but as far as I can tell when printing the Name of the objects there it seems to be an empty string. Could this be another issue with the loading function?

**Comment by dougm** [comment]
Not sure, would it be possible to share your *.xml from `govc object.save`?  That'd make it easier to reproduce.

**Comment by oniric85** [comment]
Unfortunately there are private data in it that  I can't share. Would you be able to point me to where the dc.Network is built?

I'm referring to this line https://github.com/vmware/govmomi/blob/ebcce462168193abf3ac000acdd1fe2a3dfb1b29/simulator/virtual_machine.go#L1046

Maybe I'll be able to further troubleshoot.

**Comment by dougm** [comment]
Understood.  dc.Network would be in `*-Datacenter-*.xml`:

```console
% grep network *-Datacenter-*.xml         
    <name>network</name>
      <ManagedObjectReference type="Network">network-25</ManagedObjectReference>
      <ManagedObjectReference type="Network">network-40</ManagedObjectReference>
    <name>networkFolder</name>
```

And the network names:

```console
% grep -A 1 '>name<' *-Network-*.xml
0049-Network-network-25.xml:    <name>name</name>
0049-Network-network-25.xml-    <val xmlns:XMLSchema-instance="http://www.w3.org/2001/XMLSchema-instance" XMLSchema-instance:type="xsd:string">VM Network</val>
--
0050-Network-network-40.xml:    <name>name</name>
0050-Network-network-40.xml-    <val xmlns:XMLSchema-instance="http://www.w3.org/2001/XMLSchema-instance" XMLSchema-instance:type="xsd:string">none</val>
```

**Comment by dougm** [comment]
@oniric85 fix for #1919 and the network name issue are fixed in #2170 , let us know if you have further issues
